V2

VimS opened the pull request:
Version 2 should be able to use every aspect of the DeepL-Api.
This is a work in Progress.

VimS committed:
change PHP-Version to 7.2 because it's 2020

VimS committed:
- add host to constructor
- add all 3 resources form DeepL api
- refactor buildURL-function
- add usage-fcuntion

VimS committed:
- add languages-function
- add Missiong DeepL-Parameters to buildUrl and translate functions
- refactor code for PHP7+
- Adjust Tests

VimS committed:
- fix typo

VimS committed:
composer.json:
- revert to php 5.3
- Change Author
- add ext curl und json to requirements
- remove php-duplicate from require-dev
DeepL.php
- add API_URL_SCHEMA constant
- rewrite base-url
- remove API_URL_AUTH_KEY constant
- fix some typos
- fix codestyle
- revert back to php 5.3
- add some comments
- add missin parameters to buildUrl and translate functions

VimS committed:
- add Test for languages-function
- add 2 Tests for buildBaseUrl

VimS committed:
- fix Codestyle in Tests

u-nik commented on src/DeepL.php:
```diff
@@ -157,36 +201,62 @@ public function __destruct()
-        array $tagHandling = array(),
-        array $ignoreTags = array(),
-        $formality = "default"
+        $tagHandling = null,
```
Breaking-Change: Das Argument `$tagHandling` war zuvor vom Typ Array. Im Folgenden kann aber mit einem Array nicht umgegangen werden, da in der `buildUrl()` der Wert direkt als 2. Argument an `sprintf` übergeben wird. Zuvor war da ein `implode` zu finden.

VimS replied:
Ist uns bewusst, deswegen habe ich den Branch als Version 2-alpha getagged.
Da selbst der DeepL-Simulator für den Parameter eine Entweder-oder-Auswahl vorgibt, ist hier in Zukunft nicht mit mehreren Werten für den Parameter zu rechnen, weshalb ein Array hier überflüssig ist.

Simulator:
https://www.deepl.com/docs-api/simulator/

u-nik reviewed (commented):
> @VimS habt ihr die Abhängigkeit aus dem Oxid entfernt?
> Das war nämlich damals der Grund für die PHP 5.3 Unterstützung.. :p

Hi Arkadius, du wirst es nicht glauben, aber bei uns läuft OXID bereits mit PHP 7.2, allerdings noch nicht live sondern nur im Rahmen des Neubaus aller Zahlarten. Live geht das dann im Januar. 

Aber da es keine Notwendigkeit gibt, die PHP Version anzuheben und es sich nur um ein Minor Update handelt, kann man meiner Meinung nach bei PHP 5.3 bleiben.

VimS committed:
- remove $errorCodes-Array and add Messages from DeepL-Response to DeepLException
 - fix conversion from array to url-param
 - change behaviour for $outlineDetection since DeepL-Api only needs it when it is disabled
 - fix some missing Doc-Blocks
 - Split Unit-(=local) and Integration(=api)-Tests
 - add some Testcases discovered while manual testing //I really want to refactor the Tests using Dataproviders!

VimS committed:
- check for array and key

VimS committed:
- remove buildUrl
 - remove buildBody
 - write params to post-body with http_build_query instead
 - remove Tests

VimS committed:
- remove unnecessary constants

VimS committed:
- extract buildQuery from translate

VimS committed:
- code cleanup
 - fix Bugs discovered while writing Tests
 - add Testcases for testBuildQuery

VimS committed:
- add Testcase for multiple Texts as Array
- code cleanup
- readd functionality for text as array
- try to reduce CyclomaticComplexity and NPathComplexity

VimS committed:
- further cleanup
- removeEmptyParams to reduce complexity of buildQuery
- add Testcases for removeEmptyParams

VimS committed:
- remove PHP_QUERY_RFC3986 from due to PHP 5.3

VimS committed:
- add a little documentation for a special Case

VimS committed:
- fix Docblock and Codestyle

VimS committed:
- add Testcase for Handling of html5-Tags

VimS committed:
- fix commemt for Test

VimS committed:
- remove unessesarry $resource-parameter from translate-function

VimS committed:
- fix Tests for translate()

VimS committed:
Breaking-Changes:
- Remove string-return of translate()
- Add Exception when $tagHandling is send as array instead of String
- Adjust and Add Testcases

VimS committed:
- remove checkLanguages with Tests

VimS committed:
- add Testcase for unsupported Languages

VimS committed:
- use ReflectionMethod insted of array

VimS committed:
- revert from ReflectionMethod to Array, because MessDetection did not like it.

VimS committed:
- remove unused Class

VimS committed:
Update tests/integration/DeepLApiTest.php

Co-authored-by: Dennis Exner <59438491+dexner@users.noreply.github.com>

VimS committed:
some minor improvements:
- sort class properties by type and alphabetically
- change parametes of request-Function to get rid of unnecessary variable initialization

VimS committed:
- Update README.md
- add source & target option to languages-function
- add Testscases

VimS committed:
- Updated README.md

VimS committed:
- Updated README.md

VimS committed:
- Updated README.md: User Flags of countries which are actually supported by DeepL

VimS committed:
- add Changelog.md: tried my best to add Changes done in previous Versions

VimS committed:
- add link to CHANGELOG.md in README.md

denex23 committed:
Merge pull request #17 from Baby-Markt/v2-all-post

* code check
* run unit/integration tests
* manual tests with different combinations of parameters

denex23 reviewed (approved):
* Code Check complete
* PHP Unit Tests complete
* Breaking change is documented and will be catched
* Multiple Tests with several hundred records
